test/rules/rules.bzl

```diff
@@ -0,0 +1,133 @@
+"""Generate Zserio Python library and facilitate generation of test data.
+"""
+
+load("@pip//:requirements.bzl", "requirement")
+load("@rules_python//python:defs.bzl", "py_binary", "py_library")
+
+def _dirname(f):
+    return "/".join(f.dirname.split("/")[:-1])
+
+def _impl(ctx):
+    root = _dirname(ctx.files.proto[0])
+    if root == None:
+        fail("TODO")
+
+    for s in ctx.files.deps:
+        candidate = _dirname(s)
+        if len(root) > len(candidate):
+            root = candidate
+
+    args = ctx.actions.args()
+    dir_name = ctx.label.name
+    prefix = ctx.attr.prefix
+
+    out = ctx.actions.declare_directory(dir_name)
+
+    if prefix:
+        dir_name = "{}/{}".format(dir_name, prefix)
+        args.add("-setTopLevelPackage", prefix.replace("/", "."))
+
+    args.add("-src", root)
+    args.add("-python", out.path)
+    args.add(ctx.files.proto[0].short_path.replace(root + "/", ""))
+
+    outs = [out] + [ctx.actions.declare_file(dir_name + "/" + o) for o in ctx.attr.outs]
```

Comment:
I think you need to grab the right path separator in these three places as well.

Reply:
That works for me. We can do that as part of #62